sf2cfg.1
.TH sf2cfg 1 "11 Nov 1999" \" -*-nroff-*-
.SH NAME
sf2cfg \- contruct TiMidity/KMidi cfg files for SoundFonts
.P
.SH SYNOPSIS
.B sf2cfg
sffilename [> cfgname]
.P
.SH DESCRIPTION
sf2cfg is a utility program which constructs a TiMidity configure
file, with a list instrument patch names by bank, for an AWE
SoundFont file.  Just give it the name of the SoundFont file, either
an old style .sbk font or a .sf1 or .sf2 font.  Capture the output
in a file, edit it if you want, then edit TiMidity's main
configure file (timidity.cfg) to refer to the new cfg file by
adding, e.g., lines:
.br
if 0
.br
source cfgname
.br
.SH COPYRIGHT
Copyright (C) 1998 Takashi Iwai
.SH AUTHORS
Greg Lee <lee@hawaii.edu>
.br
adapted from a program by Takashi Iwai.


ChangeLog
1999-10-21  Greg Lee  <lee@hawaii.edu>

	* resample.c: smooth off ends of patches
	  (but ocarina, vibes, pan flute still click)
	* config.h: up FRACTION_BITS to 13
	* playmidi.c: more polyphony reduction
	* kmidi.cpp: led flashes yellow when loading patches

1999-10-20  Greg Lee  <lee@hawaii.edu>

	* avoid compiler warnings
	* Makefile.am: try -O4 -Wall
	* ../configure.in.in: make it work again

1999-10-19  Greg Lee  <lee@hawaii.edu>

	* sample vibrato and modulation delay
	* kmidi.cpp: remember states of effects and filter buttons;
	  leds for chorus and reverb

1999-10-18  Greg Lee  <lee@hawaii.edu>

	* implement lfo freq modulation
	* unload least recently used patches

1999-10-16  Greg Lee  <lee@hawaii.edu>

	* kmidi.cpp: move leds higher; fix meter sync
	* sndfont.c: revise vibrato/tremolo calculation;
	  fix long delays (e.g. soprano sax)
	* resample.c: modulate pitch except for drums

1999-10-15  Greg Lee  <lee@hawaii.edu>

	* sndfont.c/playmidi.c: implement pitch effect
	  on volume envelope and modulation envelope
	  on pitch (latter only for inst. with vibrato)
	* kmidi.cpp/ctl.c,h: leds

1999-10-13  Greg Lee  <lee@hawaii.edu>

	* resample.c: portamento was continuing forever;
	  found typo in #define
	* mix.c: finish mixing freed voices faster
	* instrum.c,fffload.c: zero new sample parameters
	* kmidi.c: delay starting up meter and info windows

1999-10-12  Greg Lee  <lee@hawaii.edu>

	* resample.c: on key-up play tail of patch

1999-10-11  Greg Lee  <lee@hawaii.edu>

	* sndfont.c/resample.c: modulate cutoff filter
	* resample.c: cast right-shifted number to signed

1999-10-09  Greg Lee  <lee@hawaii.edu>

	* resample.c: cutoff filter is faster

1999-10-08  Greg Lee  <lee@hawaii.edu>

	* resample.c: change cutoff filter - ignore resonance

1999-10-07  Greg Lee  <lee@hawaii.edu>

	* sndfont.c: ad hoc envelope mode per program
	* ctl.c: kill child on exit
	* working on lyrics

1999-10-06  Greg Lee  <lee@hawaii.edu>

	* log.cpp/.h: clean up code
	* fixes from Masanao Izumo
	* playmidi.c: don't display text if have lyrics
	* sndfont.c: shorten loading messages

1999-10-04  Greg Lee  <lee@hawaii.edu>

	* playmidi.c: implement octave and fifth XG variation banks;
	  better control of dropouts
	* timidity.1: note new features and authors
	* kmidi.desktop: icon and doc names

1999-10-03  Greg Lee  <lee@hawaii.edu>

	* b_out.c: forgot to reset out_bytes after purge
	* playmidi.c,readmidi.c: some interpretation of XG variation banks
	* playmidi.c: fix volume envelope changes, allocation of
	  stereo voices
	* mix.c: fix from TiMidity++ for notes with very low volumes
	* playlist.cpp: fix "QGArray" out of bounds message
	* sndfont.c: set modes according to SF2 spec

1999-10-02  Greg Lee  <lee@hawaii.edu>

	* kmidi.cpp: filter on/off button
	* work on filter and voice reduction

1999-10-01  Greg Lee  <lee@hawaii.edu>

	* linux_a.c,b_out.c: fix purge_output, redo
	  get_outcount for meter sync
	* resample.c: low pass filter
	* playmidi.c: implement midi brightness, harmonics;
	  voice reduction
	* config.h: cspline interpolation

1999-09-30  Greg Lee  <lee@hawaii.edu>

	* kmidi.cpp: percussion decays faster on meter
	* output.c,playmidi.c: count clipped samples;
	  lower volume
	* b_out.c: non-blocking output works now
	* b_out.c,playmidi.c: try prevent dropouts
	* sndfont.c: filter percussion patches

1999-09-29  Greg Lee  <lee@hawaii.edu>

	* sndfont.c: don't load velocity ranges from
	  different fonts

1999-09-28  Greg Lee  <lee@hawaii.edu>

	* sndfont.c: fix compound instruments with
	  keyranges in presets

1999-09-27  Greg Lee  <lee@hawaii.edu>

	* kmidi.cpp: make info and meter windows independent

1999-09-26  Greg Lee  <lee@hawaii.edu>

	* config/piano+guitar,drums: more substitution
	* playlist.cpp: "select all" edit command
	* kmidi.cpp, log.cpp: put info window in panel
	* playmidi.c: fix reverb bug
	* timidity.c: note new options in help message
	* playmidi.c: show lyrics even with quiet option
	* sbk.h, sbktext.c: terminate name strings

1999-09-20  Greg Lee  <lee@hawaii.edu>

	* sf2 volume envelopes
	* sf2 peak envelope
	* "count" variables to unsigned
	* sf2 vel ranges of stereo patches

1999-09-17  Greg Lee  <lee@hawaii.edu>

	* sf2 volume envelopes (not quite right yet)

1999-09-14  Greg Lee  <lee@hawaii.edu>

	* sf2 velocity layers

1999-09-11  Greg Lee  <lee@hawaii.edu>

	* tables.c: GS volumes from TiMidity++
	* resample.c: fix core dumps
	* playmidi.c: implement exclusive instruments
	* sndfont.c: volume & pitch computations

1999-09-08  Greg Lee  <lee@hawaii.edu>

	* remove config/sndfont.cfg: obsolete
	* kmidi.cpp: tweak meter display
	* *_c.c: update other interfaces to merge extra channels
	* sndfont.c: enable stereo patch support

1999-09-06  Greg Lee  <lee@hawaii.edu>

	* up 16 channels to 64 (32 on meter display)
	* effects.c: obey echo_delay
	* timidity.c: scan over TiMidity++ "map" statements

1999-08-27  Greg Lee  <lee@hawaii.edu>

	* kmidi.cpp: check dropped file is readable midi file

1999-08-26  Greg Lee  <lee@hawaii.edu>

	* config.h: up controls/sec, FRACTINT, MAX_VOICES
	* configdlg.cpp/h: use KColorButton
	* kmidi.desktop: I play karaoke
	* kmidi.cpp/h: accept drops

1999-08-23  Greg Lee  <lee@hawaii.edu>

	* install sample midis in appdata
	* add spinbox to sync channel meter

1999-08-22  Greg Lee  <lee@hawaii.edu>

	* panel control of stereo/reverb/chorus voice & verbosity
	* midi channel pressure
	* log.cpp: set text window to read-only (no special reason)

1999-08-21  Greg Lee  <lee@hawaii.edu>

	* spinbox for polyphony; meter colors configurable

1999-08-15  Greg Lee  <lee@hawaii.edu>

	* configdlg.cpp/h, kmidi.cpp: redo to use KTabCtl, KAboutWidget
	* kmidi.cpp: add select song Combobox to KMidi panel
	* patch/ and config/ : better patches, sample timidity.cfg
	  illustrates patchset selection from front panel
	* update README, Changelog
	* version.h: 1.3-alpha
	* README.sf, CHANGES-ADAGIO: deleted

1999-08-08  Greg Lee  <lee@hawaii.edu>


sf2cfg.cpp
/*================================================================
 * sbk2cfg  --  extracts info from sf2/sbk font and constructs
 *		a TiMidity cfg file.  Greg Lee, lee@hawaii.edu, 5/98.
 * The code is adapted from "sbktext" by Takashi Iwai, which contained
 * the following notice:
 *================================================================
 * Copyright (C) 1996,1997 Takashi Iwai
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *================================================================*/

#include <stdio.h>
#include <string.h>
#include "config.h"
#include "sbk.h"
#include "sflayer.h"

static SFInfo sfinfo;

static char *getname(char *p);
static void print_sbk(SFInfo *sf, FILE *fout);

int main(int argc, char **argv)
{
	FILE *fp;

	if (argc < 2) {
		fprintf(stderr, "usage: sf2cfg SoundFontFile [> textfile]\n");
		exit(1);
	}

	if ((fp = fopen(argv[1], "r")) == NULL) {
		fprintf(stderr, "can't open SoundFont file %s\n", argv[1]);
		exit(1);
	}

	load_sbk(fp, &sfinfo);
	fclose(fp);

	fp = stdout;
	if (argc >= 3 && (fp = fopen(argv[2], "w")) == NULL) {
		fprintf(stderr, "can't open text file %s\n", argv[2]);
		exit(1);
	}
	print_sbk(&sfinfo, fp);

	fprintf(fp, "\nsf %s\n", argv[1]);

	return 0;
}


static char *getname(char *p)
{
	int i;
	static char buf[21];
	strncpy(buf, p, 20);
	buf[20] = 0;
	for (i = 19; i > 4 && buf[i]==' '; i--) {
	  buf[i] = 0;
	}
	for (i = 0; buf[i]; i++) {
	  if (buf[i] == ' ') buf[i] = '_';
	  if (buf[i] == '#') buf[i] = '@';
	}
	return buf;
}


static void print_sbk(SFInfo *sf, FILE *fout)
{
    int i, bank, preset, lastpatch;
    int lastbank = -1;
    tpresethdr *ip;
    tinsthdr *tp;

    fprintf(fout, "#  ** SoundFont: %d %d\n", sf->version, sf->minorversion);
    fprintf(fout, "#  ** SampleData: %d %d\n", (int)sf->samplepos, (int)sf->samplesize);

    fprintf(fout, "#\n#  ** Presets: %d\n", sf->nrpresets-1);

    for (bank = 0; bank <= 128; bank++) {
      for (preset = 0; preset <= 127; preset++) {
      lastpatch = -1;
	for (i = 0, ip = sf->presethdr; i < sf->nrpresets-1; i++) {
	    int b, g, inst, sm_idx;
	    if (ip[i].bank != bank) continue;
	    if (ip[i].preset != preset) continue;
	    inst = -1;
	    for (b = ip[i].bagNdx; b < ip[i+1].bagNdx; b++) {
		for (g = sf->presetbag[b]; g < sf->presetbag[b+1]; g++)
		   if (sf->presetgen[g].oper == SF_instrument) {
			inst = sf->presetgen[g].amount;
			break;
		   }
	    }
	    if (inst < 0) continue;
	    tp = sf->insthdr;
	    if (bank != 128) {
		int realwaves = 0;
		if (bank != lastbank) {
			fprintf(fout, "\nbank %d sf\n", bank);
			lastbank = bank;
		}
		for (b = tp[inst].bagNdx; b < tp[inst+1].bagNdx; b++) {
		    sm_idx = -1;
		    for (g = sf->instbag[b]; g < sf->instbag[b+1]; g++) {
			if (sf->instgen[g].oper == SF_sampleId) sm_idx = sf->instgen[g].amount;
		    }
		    if (sm_idx >= 0 && sf->sampleinfo[sm_idx].sampletype < 0x8000) realwaves++;
		}
		if (realwaves && ip[i].preset != lastpatch) {
		    fprintf(fout, "\t%3d %s\n", ip[i].preset, getname(ip[i].name));
		    lastpatch = ip[i].preset;
		}
	    }
	    else {
		int keynote, c, dpreset;
		fprintf(fout, "\ndrumset %d sf\t%s\n", ip[i].preset, getname(ip[i].name));

		for (dpreset = 0; dpreset < 128; dpreset++)
	        for (c = ip[i].bagNdx; c < ip[i+1].bagNdx; c++) {
		  inst = -1;
		  for (g = sf->presetbag[c]; g < sf->presetbag[c+1]; g++)
		   if (sf->presetgen[g].oper == SF_instrument) {
			inst = sf->presetgen[g].amount;
			break;
		   }
		  if (inst >= 0) for (b = tp[inst].bagNdx; b < tp[inst+1].bagNdx; b++) {
		    int hikeynote = -1;
		    sm_idx = keynote = -1;
		    for (g = sf->instbag[b]; g < sf->instbag[b+1]; g++) {
			if (sf->instgen[g].oper == SF_sampleId) sm_idx = sf->instgen[g].amount;
			else if (sf->instgen[g].oper == SF_keyRange) {
			    keynote = sf->instgen[g].amount & 0xff;
			    hikeynote = (sf->instgen[g].amount >> 8) & 0xff;
			}
		    }
		    if (sm_idx < 0) continue;
		    if (sf->sampleinfo[sm_idx].sampletype >= 0x8000) continue;
		    /*if (keynote != dpreset) continue;*/
		    if (dpreset < keynote) continue;
		    if (dpreset > hikeynote) continue;
		    if (sm_idx >= 0 && keynote >= 0 && dpreset != lastpatch) {
			   fprintf(fout, "\t%3d %s\n", dpreset, getname( sf->samplenames[sm_idx].name ));
			   lastpatch = dpreset;
		    }
		  }
	        }

	    }
	}
      }
    }
}
